python/ray/_private/utils.py

```diff
@@ -1187,7 +1187,17 @@ def import_attr(full_path: str):
         module_name = full_path[:last_period_idx]
         attr_name = full_path[last_period_idx + 1 :]
 
+    # `ray/dashboard` took the import precedence and can collide user's modules.
+    # Dropping `ray/dashboard` from sys.path to avoid the import collision.
+    old_paths = sys.path
+    for path in old_paths:
+        if path.endswith("ray/dashboard"):
+            sys.path.remove(path)
+
     module = importlib.import_module(module_name)
+
+    # Restore sys.path with old paths.
+    sys.path = old_paths
```

Comment:
this is sketchy, what happens if the `import_module` fails?

Reply:
Oh good point, I can wrap a try-except around this just in case. And yes will add some test cases for this. 

I don't think there are other side affect for this since we only skip dashboard when loading user's modules. Other places should all work like before. As long as `sys.path` is put back, everything else should not be affected. Even without the try-except, things should just throw exception and fail at `import_module` and nothing should be affected.

I think there are multiple places that's loading user's modules. Dashboard might just be the entrypoint when we run `serve deploy` commands. I think [replica](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/replica.py#L102) might be loading it again as well (and is compatible with changing the path, while adding `__init__.py` won't work in here). But the error we observed is solely from the dashboard. This is the `sys.path` when we first get the error running `serve deploy`
```
'/home/ray/default',
 '/home/ray/anaconda3/lib/python3.9/site-packages/ray/dashboard',
 '/home/ray/anaconda3/lib/python3.9/site-packages/ray/thirdparty_files',
 '/home/ray/anaconda3/lib/python3.9/site-packages/ray/_private/workers',
 '/tmp/ray/session_2023-06-14_11-07-23_614772_2511/runtime_resources/working_dir_files/gs_anyscale-public-cloud_org_7c1Kalm9WcX2bNIjW53GUT_cld_6zjdsbn3kbphvk3luhyrd3eg6e_workspace_snapshots_expwrk_puue3sf6399rxr6c8ustykgcyk_services__anyscale_pkg_67c9067724b5ccfddc5bc07d38e2581f',
 '/home/ray/anaconda3/lib/python39.zip',
 '/home/ray/anaconda3/lib/python3.9',
 '/home/ray/anaconda3/lib/python3.9/lib-dynload',
 '/home/ray/anaconda3/lib/python3.9/site-packages',
 '/tmp/anyscale/ray_container']
``` 
I'm guess we can probably remove all paths before `working_dir_files` as well just in case others can cause the same issue. Will address this as well.

Reply:
hmm actually if we just drop everything before `working_dir_files`, then it fails again in the replica. Will revert back to only skipping dashboard.

Reply:
Still WIP testing the behavior. Somehow it still failed with the latest changes 😅